File: frontend/src/api/unifiedTasks.ts
/**
 * 统一任务中心 API
 *
 * 对接后端 /api/v2/tasks/* 端点
 * 支持所有类型的分析任务（股票分析、持仓分析、交易复盘等）
 */

import { ApiClient } from './request'

// ==================== 类型定义 ====================

/**
 * 任务类型枚举
 */
export enum TaskType {
  STOCK_ANALYSIS = 'stock_analysis',
  POSITION_ANALYSIS = 'position_analysis',
  TRADE_REVIEW = 'trade_review',
  PORTFOLIO_HEALTH = 'portfolio_health',
  RISK_ASSESSMENT = 'risk_assessment',
  MARKET_OVERVIEW = 'market_overview',
  SECTOR_ANALYSIS = 'sector_analysis'
}

/**
 * 任务状态枚举
 */
export enum TaskStatus {
  PENDING = 'pending',
  PROCESSING = 'processing',
  COMPLETED = 'completed',
  FAILED = 'failed',
  CANCELLED = 'cancelled',
  SUSPENDED = 'suspended'  // 挂起状态（服务重启后未完成的任务）
}

/**
 * 任务列表项
 */
export interface TaskListItem {
  task_id: string
  task_type: TaskType
  status: TaskStatus
  progress: number
  created_at: string
  started_at?: string
  completed_at?: string
  execution_time: number
  error_message?: string
  
  // 任务参数（用于显示）
  symbol?: string
  code?: string
  market?: string
}

/**
 * 任务详情
 */
export interface TaskDetail {
  task_id: string
  user_id: string
  task_type: TaskType
  task_params: Record<string, any>
  workflow_id?: string
  engine_type: string
  preference_type: string
  status: TaskStatus
  progress: number
  current_step?: string
  result?: Record<string, any>
  created_at: string
  started_at?: string
  completed_at?: string
  execution_time: number
  error_message?: string
  retry_count: number
  max_retries: number
  tokens_used: number
  cost: number
}

/**
 * 任务统计
 */
export interface TaskStatistics {
  total: number
  pending: number
  processing: number
  completed: number
  failed: number
  cancelled: number
}

/**
 * 任务列表查询参数
 */
export interface TaskListParams {
  task_type?: TaskType
  status?: TaskStatus
  limit?: number
  skip?: number
}

// ==================== API 方法 ====================

/**
 * 获取任务列表
 */
export function getTaskList(params?: TaskListParams) {
  return ApiClient.get<{
    tasks: TaskListItem[]
    total: number
    limit: number
    skip: number
  }>('/api/v2/tasks/list', params)
}

/**
 * 获取任务统计
 */
export function getTaskStatistics() {
  return ApiClient.get<TaskStatistics>('/api/v2/tasks/statistics')
}

/**
 * 获取任务详情
 */
export function getTaskDetail(taskId: string) {
  return ApiClient.get<TaskDetail>(`/api/v2/tasks/${taskId}`)
}

/**
 * 取消任务
 */
export function cancelTask(taskId: string) {
  return ApiClient.delete<{ success: boolean; message: string }>(`/api/v2/tasks/${taskId}`)
}

/**
 * 恢复挂起的任务
 */
export function resumeTask(taskId: string) {
  return ApiClient.post<{ success: boolean; message: string }>(`/api/v2/tasks/${taskId}/resume`)
}

/**
 * 任务类型中文名称映射
 */
export const TaskTypeNames: Record<TaskType, string> = {
  [TaskType.STOCK_ANALYSIS]: '股票分析',
  [TaskType.POSITION_ANALYSIS]: '持仓分析',
  [TaskType.TRADE_REVIEW]: '交易复盘',
  [TaskType.PORTFOLIO_HEALTH]: '组合健康度',
  [TaskType.RISK_ASSESSMENT]: '风险评估',
  [TaskType.MARKET_OVERVIEW]: '市场概览',
  [TaskType.SECTOR_ANALYSIS]: '板块分析'
}

/**
 * 任务状态中文名称映射
 */
export const TaskStatusNames: Record<TaskStatus, string> = {
  [TaskStatus.PENDING]: '等待中',
  [TaskStatus.PROCESSING]: '进行中',
  [TaskStatus.COMPLETED]: '已完成',
  [TaskStatus.FAILED]: '失败',
  [TaskStatus.CANCELLED]: '已取消',
  [TaskStatus.SUSPENDED]: '已挂起'
}

/**
 * 任务状态颜色映射
 */
export const TaskStatusColors: Record<TaskStatus, string> = {
  [TaskStatus.PENDING]: 'info',
  [TaskStatus.PROCESSING]: 'warning',
  [TaskStatus.COMPLETED]: 'success',
  [TaskStatus.FAILED]: 'danger',
  [TaskStatus.CANCELLED]: 'info',
  [TaskStatus.SUSPENDED]: 'warning'
}

File: app/routers/unified_tasks.py
"""
统一任务中心 API

提供统一的任务查询接口，支持所有类型的分析任务
"""

from typing import List, Optional
from fastapi import APIRouter, HTTPException, Depends, Query
from pydantic import BaseModel
import logging

from app.models.analysis import AnalysisTaskType, AnalysisStatus
from app.models.user import PyObjectId
from app.services.task_analysis_service import get_task_analysis_service
from app.routers.auth_db import get_current_user

logger = logging.getLogger(__name__)

router = APIRouter(prefix="/api/v2/tasks", tags=["unified-task-center"])


# ==================== 响应模型 ====================

class TaskListItem(BaseModel):
    """任务列表项"""
    task_id: str
    task_type: str
    status: str
    progress: int
    created_at: str
    started_at: Optional[str] = None
    completed_at: Optional[str] = None
    execution_time: float = 0.0
    error_message: Optional[str] = None
    
    # 任务参数（用于显示）
    symbol: Optional[str] = None
    code: Optional[str] = None
    market: Optional[str] = None


class TaskStatistics(BaseModel):
    """任务统计"""
    total: int
    pending: int
    processing: int
    completed: int
    failed: int
    cancelled: int


# ==================== API 端点 ====================

@router.get("/list", summary="获取任务列表")
async def get_task_list(
    task_type: Optional[AnalysisTaskType] = Query(None, description="任务类型过滤"),
    status: Optional[AnalysisStatus] = Query(None, description="状态过滤"),
    limit: Optional[int] = Query(None, description="返回数量（可选，不设置则返回所有）"),
    skip: int = Query(0, ge=0, description="跳过数量"),
    user: dict = Depends(get_current_user)
):
    """获取用户的任务列表

    支持按任务类型和状态过滤

    注意：如果不指定 limit，将返回所有记录，由前端处理分页
    """
    try:
        logger.info(f"🎯 [统一任务中心] 获取任务列表请求")
        logger.info(f"👤 用户信息: {user}")
        logger.info(f"📊 查询参数: task_type={task_type}, status={status}, limit={limit}, skip={skip}")

        user_id = PyObjectId(user["id"])
        logger.info(f"🔄 转换后的user_id: {user_id} (类型: {type(user_id)})")

        service = get_task_analysis_service()

        # 如果没有指定 limit，则返回所有记录
        actual_limit = limit if limit is not None else 999999

        tasks = await service.list_user_tasks(
            user_id=user_id,
            task_type=task_type,
            status=status,
            limit=actual_limit,
            skip=skip
        )
        
        # 转换为列表项格式
        items = []
        for task in tasks:
            # 从任务参数中提取显示信息
            symbol = task.task_params.get("symbol")
            code = task.task_params.get("code")
            market = task.task_params.get("market") or task.task_params.get("market_type")
            
            items.append(TaskListItem(
                task_id=task.task_id,
                task_type=task.task_type,
                status=task.status,
                progress=task.progress,
                created_at=task.created_at.isoformat(),
                started_at=task.started_at.isoformat() if task.started_at else None,
                completed_at=task.completed_at.isoformat() if task.completed_at else None,
                execution_time=task.execution_time,
                error_message=task.error_message,
                symbol=symbol,
                code=code,
                market=market
            ))
        
        return {
            "success": True,
            "data": {
                "tasks": items,
                "total": len(items),
                "limit": limit,
                "skip": skip
            },
            "message": "ok"
        }
        
    except Exception as e:
        logger.error(f"获取任务列表失败: {e}")
        raise HTTPException(status_code=500, detail=f"获取任务列表失败: {str(e)}")


@router.get("/statistics", summary="获取任务统计")
async def get_task_statistics(
    user: dict = Depends(get_current_user)
):
    """获取用户的任务统计信息"""
    try:
        logger.info(f"🎯 [统一任务中心] 获取任务统计请求")
        logger.info(f"👤 用户信息: {user}")

        user_id = PyObjectId(user["id"])
        logger.info(f"🔄 转换后的user_id: {user_id} (类型: {type(user_id)})")

        service = get_task_analysis_service()

        stats = await service.get_task_statistics(user_id)
        logger.info(f"📊 统计结果: {stats}")
        
        return {
            "success": True,
            "data": TaskStatistics(**stats),
            "message": "ok"
        }
        
    except Exception as e:
        logger.error(f"获取任务统计失败: {e}")
        raise HTTPException(status_code=500, detail=f"获取任务统计失败: {str(e)}")


@router.get("/{task_id}", summary="获取任务详情")
async def get_task_detail(
    task_id: str,
    user: dict = Depends(get_current_user)
):
    """获取任务详情"""
    try:
        user_id = PyObjectId(user["id"])
        service = get_task_analysis_service()

        task = await service.get_task(task_id)
        if not task:
            raise HTTPException(status_code=404, detail="任务不存在")

        # 验证权限
        if task.user_id != user_id:
            raise HTTPException(status_code=403, detail="无权访问此任务")
        
        return {
            "success": True,
            "data": task.model_dump(),
            "message": "ok"
        }
        
    except HTTPException:
        raise
    except Exception as e:
        logger.error(f"获取任务详情失败: {e}")
        raise HTTPException(status_code=500, detail=f"获取任务详情失败: {str(e)}")


@router.delete("/{task_id}", summary="取消任务")
async def cancel_task(
    task_id: str,
    user: dict = Depends(get_current_user)
):
    """取消任务"""
    try:
        user_id = PyObjectId(user["id"])
        service = get_task_analysis_service()

        # 验证任务存在和权限
        task = await service.get_task(task_id)
        if not task:
            raise HTTPException(status_code=404, detail="任务不存在")

        if task.user_id != user_id:
            raise HTTPException(status_code=403, detail="无权访问此任务")

        # 取消任务
        success = await service.cancel_task(task_id)

        if success:
            return {
                "success": True,
                "message": "任务已取消"
            }
        else:
            return {
                "success": False,
                "message": "任务无法取消（可能已完成或失败）"
            }

    except HTTPException:
        raise
    except Exception as e:
        logger.error(f"取消任务失败: {e}")
        raise HTTPException(status_code=500, detail=f"取消任务失败: {str(e)}")


@router.post("/{task_id}/resume", summary="恢复挂起的任务")
async def resume_task(
    task_id: str,
    user: dict = Depends(get_current_user)
):
    """恢复挂起的任务，将任务重新加入队列"""
    try:
        user_id = PyObjectId(user["id"])
        service = get_task_analysis_service()

        # 验证任务存在和权限
        task = await service.get_task(task_id)
        if not task:
            raise HTTPException(status_code=404, detail="任务不存在")

        if task.user_id != user_id:
            raise HTTPException(status_code=403, detail="无权访问此任务")

        # 检查任务状态
        if task.status != AnalysisStatus.SUSPENDED:
            raise HTTPException(status_code=400, detail=f"只能恢复挂起状态的任务，当前状态: {task.status}")

        # 恢复任务
        success = await service.resume_task(task_id)

        if success:
            return {
                "success": True,
                "message": "任务已恢复，正在重新排队..."
            }
        else:
            return {
                "success": False,
                "message": "恢复任务失败"
            }

    except HTTPException:
        raise
    except Exception as e:
        logger.error(f"恢复任务失败: {e}")
        raise HTTPException(status_code=500, detail=f"恢复任务失败: {str(e)}")

